<launch>

	<arg name="ns" default="/"/>
	<arg name="show_debug_info" default="false"/>

	<!-- control period [s] -->
	<arg name="ctrl_period" default="0.04"/>

	<!-- length span for resampling trajectory [m] -->
	<arg name="traj_resample_dist" default="0.1"/>

	 <!-- recalculate yaw angle after resampling -->
	 <arg name="enable_yaw_recalculation" default="true"/>

	<!-- stop mpc calculation when error is larger than the following value -->
	<arg name="admisible_position_error" default="5.0"/>
	<arg name="admisible_yaw_error_deg" default="90.0"/>

	<!-- path smoothing flug, must true except debugging -->
	<arg name="enable_path_smoothing" default="true"/>

	<arg name="path_smoothing_times" default="1"/>

	<!-- path smoothing low pass filler, moving average number -->
	<arg name="path_filter_moving_ave_num" default="35"/>

	<!-- curvature calculation smoothing, larger makes smoother
         curvature is calculated from three points: p(i-num), p(i), p(i+num) -->
	<arg name="curvature_smoothing_num" default="35"/>

	<!-- low pass filter for steering output command [Hz] -->
	<arg name="steering_lpf_cutoff_hz" default="2.0"/>

	<!-- optimization solver type. option is "unconstraint_fast", "unconstraint", and "qpoases_hotstart" -->
	<arg name="mpc_qp_solver_type" default="unconstraint_fast"/>

	<!-- vehicle model type for mpc prediction. option is "kinematics", "kinematics_no_delay" -->
	<arg name="mpc_vehicle_model_type" default="kinematics"/>

	<!-- total prediction step -->
	<arg name="mpc_n" default="70"/>

	<!-- prediction period for one step [s] (predict holizon period = npc_n * mpc_dt) -->
	<arg name="mpc_dt" default="0.1"/>

	<!-- for Q lateral error -->
	<arg name="mpc_weight_lat_error" default="0.1"/>

	<!-- for Q heading error -->
	<arg name="mpc_weight_heading_error" default="0.0"/>

	<!-- or Q heading error * vel -->
	<arg name="mpc_weight_heading_error_squared_vel_coeff" default="0.3"/>

	<!-- for R referencve input error -->
	<arg name="mpc_weight_steering_input" default="1.0"/>

	<!-- for R referencve input error * vel -->
	<arg name="mpc_weight_steering_input_squared_vel_coeff" default="0.25"/>

	<!-- to reduce lateral jerk -->
	<arg name="mpc_weight_lat_jerk" default="0.0"/>

	<!-- endpoint error weight to improve mpc stability -->
	<arg name="mpc_weight_endpoint_Q_scale" default="5.0"/>

	<!-- To reduce path noise, feedforward angle is set to zero when it is smaller than this value. -->
	<arg name="mpc_zero_ff_steer_deg" default="2.0"/>

	<!-- steering dynamics time constant (1d approzimation) [s] -->
	<arg name="vehicle_model_steer_tau" default="0.3"/>

	<!-- wheel base length [m] -->
	<arg name="vehicle_model_wheelbase" default="2.9"/>

	<!-- steering angle limit [deg] -->
	<arg name="steer_lim_deg" default="40.0"/>




	<group ns="$(arg ns)">

		<node pkg="waypoint_follower" type="mpc_waypoints_converter" name="mpc_waypoints_converter" output="screen"/>

		<node pkg="waypoint_follower" type="mpc_follower" name="mpc_follower" output="screen">
			<!-- system -->
			<param name="publish_debug_values" value="true"/>
			<param name="ctrl_cmd_interface" value="all"/>
			<param name="in_waypoints_name" value="/mpc_waypoints"/>
			<param name="in_vehicle_status_name" value="/sim_vehicle_status"/>
			<param name="in_selfpose_name" value="/current_pose"/>
			<param name="out_twist_name" value="/twist_raw"/>
			<param name="out_vehicle_cmd_name" value="/ctrl_cmd"/>

			<!-- paramters -->
			<param name="ctrl_period" value="$(arg ctrl_period)"/>
			<param name="traj_resample_dist" value="$(arg traj_resample_dist)"/>
			<param name="admisible_position_error" value="$(arg admisible_position_error)"/>
			<param name="admisible_yaw_error_deg" value="$(arg admisible_yaw_error_deg)"/>
			<param name="path_smoothing_times" value="$(arg path_smoothing_times)"/>


			<param name="show_debug_info" value="$(arg show_debug_info)"/>
			<param name="enable_yaw_recalculation" value="$(arg enable_yaw_recalculation)"/>
			<param name="enable_path_smoothing" value="$(arg enable_path_smoothing)"/>
			<param name="path_filter_moving_ave_num" value="$(arg path_filter_moving_ave_num)"/>
			<param name="curvature_smoothing_num" value="$(arg curvature_smoothing_num)"/>
			<param name="steering_lpf_cutoff_hz" value="$(arg steering_lpf_cutoff_hz)"/>
			<param name="qp_solver_type" value="$(arg mpc_qp_solver_type)"/>
			<param name="vehicle_model_type" value="$(arg mpc_vehicle_model_type)"/>
			<param name="mpc_n" value="$(arg mpc_n)"/>
			<param name="mpc_dt" value="$(arg mpc_dt)"/>
			<param name="mpc_weight_lat_error" value="$(arg mpc_weight_lat_error)"/>
			<param name="mpc_weight_heading_error" value="$(arg mpc_weight_heading_error)"/>
			<param name="mpc_weight_heading_error_squared_vel_coeff" value="$(arg mpc_weight_heading_error_squared_vel_coeff)"/>
			<param name="mpc_weight_steering_input" value="$(arg mpc_weight_steering_input)"/>
			<param name="mpc_weight_steering_input_squared_vel_coeff" value="$(arg mpc_weight_steering_input_squared_vel_coeff)"/>
			<param name="mpc_weight_lat_jerk" value="$(arg mpc_weight_lat_jerk)"/>
			<param name="mpc_weight_endpoint_Q_scale" value="$(arg mpc_weight_endpoint_Q_scale)"/>
			<param name="mpc_zero_ff_steer_deg" value="$(arg mpc_zero_ff_steer_deg)"/>
			<param name="vehicle_model_steer_tau" value="$(arg vehicle_model_steer_tau)"/>
			<param name="vehicle_model_wheelbase" value="$(arg vehicle_model_wheelbase)"/>
			<param name="steer_lim_deg" value="$(arg steer_lim_deg)"/>
		</node>

	</group>

</launch>
